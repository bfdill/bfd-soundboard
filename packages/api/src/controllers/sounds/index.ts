import * as KoaRouter from 'koa-router'
import { NOT_FOUND, OK } from 'http-status-codes'
import { ISoundService, SoundService } from '@talkboy-9000/sounds'

import { IApplicationContext, ApplicationState } from '../../types'

export interface ISoundsController {
  get: (ctx: IApplicationContext) => void
}

export class SoundsController implements ISoundsController {
  private static instance: ISoundsController | undefined

  constructor(protected readonly soundService: ISoundService) {}

  get = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'SoundsController',
        method: 'get'
      }
    })
    const sounds = this.soundService.getSounds(logger)

    context.body = { sounds }

    if (sounds === null) {
      logger.warn('no sounds from service')
      context.status = NOT_FOUND
      return
    }

    logger.info({
      soundsLength: sounds.length
    })
    context.status = OK
  }

  static getInstance(): ISoundsController {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundsController(SoundService.getInstance())

    return this.instance
  }
}

export const getSoundsRouter = () =>
  new KoaRouter<ApplicationState, IApplicationContext>().get(
    '/',
    SoundsController.getInstance().get
  )
