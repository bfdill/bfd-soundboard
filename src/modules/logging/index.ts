import * as winston from 'winston'

export const transports = [
  new winston.transports.Console({
    level: 'silly',
    format: winston.format.prettyPrint()
  })
]

export const logger = winston.createLogger({
  transports,
  defaultMeta: {
    module: 'default'
  }
})

export const createModuleLogger = (module: string) =>
  winston.createLogger({
    transports,
    defaultMeta: { module },
    levels: winston.config.npm.levels
  })
