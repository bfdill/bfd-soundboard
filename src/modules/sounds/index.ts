import { join, resolve, sep, basename } from 'path'
import * as winston from 'winston'
import * as sane from 'sane'
import { sync } from 'glob'
import { getAppLogger } from '../../app'

export const PATH_TO_SOUNDS = join(process.cwd(), 'audio')

export type Sound = {
  id: string
  filename: string
  basename: string
}

export interface ISoundService {
  getBySoundId(soundId: string, parentLogger: winston.Logger): Sound | undefined
  getSounds: (parentLogger: winston.Logger) => Sound[]
  isPathValid: (filename: string, parentLogger: winston.Logger) => boolean
}

export class SoundService implements ISoundService {
  private static instance: ISoundService | undefined
  public readonly FILE_GLOB = '**/*.mp3'

  protected sounds: Sound[] = []
  protected readonly watcher: sane.Watcher

  constructor(
    public readonly pathToSounds: string,
    protected readonly saneFunction: typeof sane,
    protected readonly logger: winston.Logger
  ) {
    this.addSounds()
    this.watcher = this.createWatch()
  }

  addSounds = () => {
    this.logger.info('add sounds')

    sync(join(this.pathToSounds, this.FILE_GLOB)).forEach(this.addSound)
  }

  addSound = (filename: string) => {
    this.logger.debug({ filename })

    this.sounds = [
      ...this.sounds,
      {
        filename,
        id: filename,
        basename: basename(filename)
      }
    ]
      .sort((a, b) => a.filename.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      filename,
      sounds: this.sounds
    })
  }

  createWatch = (): sane.Watcher => {
    return this.saneFunction(this.pathToSounds, { glob: this.FILE_GLOB })
      .on('add', (path: string) => {
        this.addSound(path)

        this.logger.debug({
          path,
          message: 'watch:add'
        })
      })
      .on('delete', (path: string) => {
        this.removeSound(path)

        this.logger.debug({
          path,
          message: 'watch:delete'
        })
      })
  }

  getBySoundId = (
    soundId: string,
    parentLogger: winston.Logger
  ): Sound | undefined => {
    const result = this.sounds.find(s => s.id === soundId)
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getBySoundId'
        }
      })
      .debug({
        result,
        soundId,
        message: `getBySoundId(${soundId})`
      })
    return result
  }

  getSounds = (parentLogger: winston.Logger): Sound[] => {
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getSounds'
        }
      })
      .silly({ sounds: this.sounds })
    return this.sounds
  }

  isPathValid = (filename: string, parentLogger: winston.Logger): boolean => {
    const absCandidate = resolve(filename) + sep
    const result =
      absCandidate.substring(0, PATH_TO_SOUNDS.length) === PATH_TO_SOUNDS
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'isPathValid'
        }
      })
      .debug({
        filename,
        result,
        message: `isPathValid(${filename})`,
        pathToSounds: PATH_TO_SOUNDS
      })
    return result
  }

  removeSound = (filename: string) => {
    this.logger.debug(`removeSound(${filename})`)

    const soundIndex = this.sounds.findIndex(s => s.filename === filename)

    if (soundIndex === -1) return

    this.sounds = this.sounds
      .slice(0, soundIndex)
      .concat(this.sounds.slice(soundIndex + 1))
      .sort((a, b) => `${a.filename}`.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      message: `removeSound(${filename})`,
      sounds: this.sounds
    })
  }

  static getInstance(): ISoundService {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundService(
      PATH_TO_SOUNDS,
      sane,
      getAppLogger().child({
        service: {
          name: 'SoundService'
        }
      })
    )

    return this.instance
  }
}
