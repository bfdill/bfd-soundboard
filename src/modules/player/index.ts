// import * as playSound from 'play-sound'
import * as winston from 'winston'
import { ApplicationState } from '../../types'
import { createPlayer, IPlayer } from './types'
import { ISoundService, SoundService } from '../sounds'

// tsd creation fail
const createPlayer: createPlayer = require('play-sound')

export interface IPlayerService {
  playFile: (
    filename: string,
    state: ApplicationState,
    parentLogger: winston.Logger
  ) => PromiseLike<void>
}

export class PlayerService implements IPlayerService {
  static instance: IPlayerService | undefined

  constructor(
    protected readonly player: IPlayer,
    protected readonly soundService: ISoundService
  ) {}

  playFile = (
    filename: string,
    state: ApplicationState,
    parentLogger: winston.Logger
  ): PromiseLike<void> => {
    const logger = parentLogger.child({
      service: 'PlayerService',
      method: 'playFile'
    })

    logger.info({ filename })

    if (!this.soundService.isPathValid(filename, state, logger)) {
      const message = 'filename path is invalid'
      logger.error({ filename, message })
      return Promise.reject(message)
    }

    return new Promise((resolve, reject) => {
      const next = (error?: any) => {
        if (error === undefined || error === null) {
          logger.debug({
            filename,
            message: 'playback success'
          })
          resolve()
          return
        }

        logger.debug({
          error,
          filename,
          message: 'playback failure'
        })
        reject(error)
      }

      this.player.play(filename, next)
    })
  }

  static getInstance() {
    if (this.instance !== undefined) return this.instance

    this.instance = new PlayerService(
      createPlayer(),
      SoundService.getInstance()
    )

    return this.instance
  }
}
