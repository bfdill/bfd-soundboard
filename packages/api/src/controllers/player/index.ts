import * as koaRouter from 'koa-router'
import { NOT_FOUND, BAD_REQUEST, OK } from 'http-status-codes'
import { ISoundService, SoundService } from '../../modules/sounds'
import { IPlayerService, PlayerService } from '../../modules/player'
import { randomIntFromInterval } from '@bfd/utils'
import { ApplicationState, IApplicationContext } from '../../types'

export interface IPlayerController {
  playSound: (context: IApplicationContext) => PromiseLike<void>
  playRando: (context: IApplicationContext) => PromiseLike<void>
}

export class PlayerController implements IPlayerController {
  private static instance: IPlayerController | undefined

  constructor(
    protected readonly playerService: IPlayerService,
    protected readonly soundService: ISoundService
  ) {}

  playSound = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'PlayerController',
        method: 'playSound'
      }
    })
    const { soundId } = context.params
    logger.info({ soundId })

    if (typeof soundId !== 'string') {
      logger.error({
        soundId,
        message: 'soundId !== string'
      })
      context.body = { soundId }
      context.status = BAD_REQUEST
      return
    }

    const sound = this.soundService.getBySoundId(soundId, logger)

    if (sound === undefined) {
      logger.error({
        soundId,
        message: 'sound === undefined'
      })
      context.body = { soundId }
      context.status = NOT_FOUND
      return
    }

    logger.debug({
      sound,
      soundId,
      message: 'playFile'
    })

    await this.playerService.playFile(sound.filename, logger)

    context.body = { sound }
    context.status = OK
  }

  playRando = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'PlayerController',
        method: 'playRando'
      }
    })
    logger.info('request')
    const sounds = this.soundService.getSounds(logger)

    if (sounds.length === 0) {
      logger.error({
        message: 'sounds.length === 0'
      })
      context.status = NOT_FOUND
      return
    }

    logger.silly({
      soundLength: sounds.length,
      message: 'sounds length'
    })

    const sound = sounds[randomIntFromInterval(0, sounds.length - 1)]

    logger.debug({
      sound,
      message: 'selected sound'
    })

    await this.playerService.playFile(sound.filename, logger)

    context.body = { sound }
    context.status = OK
  }

  static getInstance(): IPlayerController {
    if (this.instance !== undefined) return this.instance

    this.instance = new PlayerController(
      PlayerService.getInstance(),
      SoundService.getInstance()
    )

    return this.instance
  }
}

export const getPlayerRouter = () =>
  new koaRouter<ApplicationState, IApplicationContext>()
    .get('/rando', PlayerController.getInstance().playRando)
    .get('/:soundId', PlayerController.getInstance().playSound)
