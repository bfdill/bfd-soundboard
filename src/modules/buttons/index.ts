import * as winston from 'winston'
import { inspect } from 'util'
import * as rpiGpio from 'rpi-gpio'
import { ISoundService, SoundService } from '../sounds'
import { IPlayerService, PlayerService } from '../player'

export type Button = {
  switch: rpiGpio.Pins
  led: rpiGpio.Pins
  soundId: string
}

export type GpioSetupConfig = {
  pin: rpiGpio.Pins
  direction: rpiGpio.Direction
  edge: rpiGpio.Edge
  cb: () => void
}

export type SwitchState = {
  active: boolean
  pushed: boolean
  button: Button
}

const buttonList: Button[] = [
  { switch: '3', led: '8', soundId: '0' },
  { switch: '5', led: '10', soundId: '1' },
  { switch: '7', led: '12', soundId: '2' },
  { switch: '11', led: '16', soundId: '3' },
  { switch: '13', led: '18', soundId: 'rando' },
  { switch: '15', led: '22', soundId: 'rando' },
  { switch: '19', led: '24', soundId: '4' },
  { switch: '21', led: '26', soundId: '5' },
  { switch: '23', led: '32', soundId: '6' },
  { switch: '29', led: '36', soundId: '7' }
]

export interface IButtonService {
  ledOn: (button: Button, logger: winston.Logger) => Promise<void>
  setupFromConfig: (config: GpioSetupConfig) => Promise<void>
  registerChangeListener: (logger: winston.Logger) => () => void
  startButtonService: (baseLogger: winston.Logger) => Promise<void>
}

export class ButtonService implements IButtonService {
  private static instance: IButtonService | undefined
  protected readonly gpiop: rpiGpio.IGpioPromise
  protected readonly registeredSwitches: {
    [key: string]: SwitchState
  } = {}

  constructor(
    protected readonly rpiGpio: rpiGpio.IGpio,
    public readonly buttons: Button[],
    protected readonly soundService: ISoundService,
    protected readonly playerService: IPlayerService
  ) {
    this.gpiop = rpiGpio.promise
  }

  ledOn = (button: Button, logger: winston.Logger): Promise<void> => {
    logger.debug({
      button,
      method: 'ledOn'
    })

    return this.gpiop.write(button.led, true)
  }

  protected static setupCallback(
    button: Button,
    pin: rpiGpio.Pins,
    logger: winston.Logger
  ) {
    return (err: any) => {
      logger.log({
        button,
        pin,
        err,
        level: err ? 'error' : 'debug',
        message: 'setupCallback'
      })

      if (err) throw err
    }
  }

  protected static ledConfigFromButton(
    button: Button,
    logger: winston.Logger
  ): GpioSetupConfig {
    return {
      pin: button.led,
      direction: 'out',
      edge: 'none',
      cb: () => this.setupCallback(button, button.led, logger)
    }
  }

  protected static switchConfigFromButton(
    button: Button,
    logger: winston.Logger
  ): GpioSetupConfig {
    return {
      pin: button.switch,
      direction: 'in',
      edge: 'both',
      cb: () => this.setupCallback(button, button.switch, logger)
    }
  }

  setupFromConfig = (config: GpioSetupConfig) =>
    this.gpiop
      .setup(config.pin, config.direction, config.edge)
      .then(() => config.cb())

  registerChangeListener = (logger: winston.Logger) => () => {
    this.rpiGpio.on('change', (channel: rpiGpio.Pins, value: boolean) => {
      const sw = this.registeredSwitches[channel]
      const pushed = !value

      if (sw === undefined || sw === null) return
      if (sw.active === pushed) return
      if (!pushed) {
        sw.active = false
        return
      }

      sw.active = true
      const fileToPlay = this.soundService.getBySoundId(
        sw.button.soundId,
        logger
      )

      if (fileToPlay === undefined) {
        throw new Error(
          `unknown fileToPlay for button(${inspect(sw, false, 5)})`
        )
      }

      this.playerService.playFile(fileToPlay.filename, logger)
    })
  }

  startButtonService = (baseLogger: winston.Logger) => {
    const logger = baseLogger.child({
      service: 'ButtonService',
      method: 'startButtonService'
    })

    return Promise.all(
      this.buttons
        .map(btn => [
          this.setupFromConfig(ButtonService.ledConfigFromButton(btn, logger)),
          this.setupFromConfig(
            ButtonService.switchConfigFromButton(btn, logger)
          )
        ])
        .flat()
    )
      .then(() => {
        this.buttons.forEach(btn => {
          this.registeredSwitches[btn.switch] = {
            active: false,
            button: btn,
            pushed: false
          }
        })
      })
      .then(() => logger.info('setup ðŸ‘œðŸ‘œðŸ‘œ complete'))
      .then(() =>
        logger.debug({
          message: 'registeredSwitches',
          registeredSwitches: this.registeredSwitches
        })
      )
      .then(this.registerChangeListener(logger))
      .catch(cause => {
        logger.error({
          cause,
          message: 'startButtonService error'
        })
      })
  }

  static getInstance(): IButtonService {
    if (this.instance !== undefined) return this.instance

    return new ButtonService(
      rpiGpio,
      buttonList,
      SoundService.getInstance(),
      PlayerService.getInstance()
    )
  }
}
