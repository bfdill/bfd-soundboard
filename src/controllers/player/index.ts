import * as koaRouter from 'koa-router'
import { NOT_FOUND, BAD_REQUEST, OK } from 'http-status-codes'
import { createModuleLogger } from '../../modules/logging'
import winston = require('winston')
import { ISoundService, soundService } from '../../modules/sounds'
import { IPlayerService, playerService } from '../../modules/player'
import { randomIntFromInterval } from '../../utils'

const playerControllerLogger = createModuleLogger('PlayerController')

export interface IPlayerController {
  playSound: (ctx: koaRouter.RouterContext) => PromiseLike<void>
  playRando: (ctx: koaRouter.RouterContext) => PromiseLike<void>
}

export class PlayerController implements IPlayerController {
  constructor(
    protected readonly playerService: IPlayerService,
    protected readonly soundService: ISoundService,
    protected readonly logger: winston.Logger
  ) {}

  playSound = async (ctx: koaRouter.RouterContext) => {
    const { soundId } = ctx.params

    if (typeof soundId !== 'string') {
      this.logger.warn({
        soundId,
        message: `playSound().soundId !== string (${soundId})`
      })
      ctx.status = BAD_REQUEST
      ctx.body = { soundId }
      return
    }

    const sound = this.soundService.getBySoundId(soundId)

    if (sound === undefined) {
      this.logger.error({
        soundId,
        message: 'playSound().sound === undefined'
      })
      ctx.status = NOT_FOUND
      ctx.body = { soundId }
      return
    }

    this.logger.debug({
      sound,
      soundId,
      message: 'playSound()'
    })
    await this.playerService.playFile(sound.filename)
    ctx.status = OK
    ctx.body = { sound }
  }

  playRando = async (ctx: koaRouter.RouterContext) => {
    this.logger.info({
      message: 'playRando()'
    })
    const sounds = this.soundService.getSounds()

    if (sounds === null) {
      this.logger.error({
        message: 'playRando().sounds === null'
      })
      ctx.status = NOT_FOUND
      return
    }

    const sound =
      sounds.length === 0
        ? sounds[0]
        : sounds[randomIntFromInterval(0, sounds.length - 1)]

    this.logger.debug({
      sound,
      message: 'playRando()'
    })
    await this.playerService.playFile(sound.filename)
    ctx.body = { sound }
  }
}

const controllerInstance = new PlayerController(
  playerService,
  soundService,
  playerControllerLogger
)

export const playerRouter = new koaRouter()
  .get('/rando', controllerInstance.playRando)
  .get('/:soundId', controllerInstance.playSound)
