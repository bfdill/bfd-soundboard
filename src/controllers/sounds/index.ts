import * as KoaRouter from 'koa-router'
import { NOT_FOUND, OK } from 'http-status-codes'

import { IApplicationContext, ApplicationState } from '../../types'
import { ISoundService, soundService } from '../../modules/sounds'

export interface ISoundsController {
  get: (ctx: IApplicationContext) => void
}

export class SoundsController implements ISoundsController {
  private static instance: ISoundsController | undefined

  constructor(protected readonly soundService: ISoundService) {}

  get = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'SoundsController',
        method: 'get'
      }
    })
    const sounds = this.soundService.getSounds()

    context.body = { sounds }

    if (sounds === null) {
      logger.warn('no sounds from service')
      context.status = NOT_FOUND
      return
    }

    logger.info({
      soundsLength: sounds.length
    })
    context.status = OK
  }

  static getInstance(): ISoundsController {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundsController(soundService)

    return this.instance
  }
}

export const getSoundsRouter = () =>
  new KoaRouter<ApplicationState, IApplicationContext>().get(
    '/',
    SoundsController.getInstance().get
  )
