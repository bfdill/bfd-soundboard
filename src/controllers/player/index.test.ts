import * as koaRouter from 'koa-router'
import { PlayerController, IPlayerController } from '.'
import { IPlayerService } from '../../modules/player'
import { ISoundService, Sound } from '../../modules/sounds'
import winston = require('winston')

describe('controllers -> player', () => {
  const contextBuilder = (soundId: any): koaRouter.RouterContext =>
    ({
      params: { soundId }
    } as any)

  test('has known exports', () => {
    expect(Object.keys(require('.'))).toMatchSnapshot()
  })

  describe('playSound', () => {
    const playerService: IPlayerService = {
      playFile: jest.fn()
    } as any
    const mockGetBySoundId = jest.fn()
    const soundService: ISoundService = {
      getBySoundId: mockGetBySoundId
    } as any
    const logger: winston.Logger = {
      debug: jest.fn(),
      error: jest.fn(),
      warn: jest.fn()
    } as any
    let playerController: IPlayerController

    beforeEach(() => {
      playerController = new PlayerController(
        playerService,
        soundService,
        logger
      )
    })

    test.each([undefined, null, false, true, 1])(
      'bad request on non-string',
      async (soundId: any) => {
        expect.assertions(2)
        const ctx = contextBuilder(soundId)

        await playerController.playSound(ctx)

        expect(ctx).toMatchSnapshot()
        expect(logger.warn).toMatchSnapshot()
      }
    )

    test('not found on unknown song', async () => {
      expect.assertions(2)

      const ctx = contextBuilder('unknown')

      await playerController.playSound(ctx)

      expect(ctx).toMatchSnapshot()
      expect(logger.error).toMatchSnapshot()
    })

    test('plays sound', async () => {
      expect.assertions(3)

      const sound: Sound = {
        basename: 'mommy shark',
        filename: 'daddy shark',
        id: 'baby_shark'
      }
      const ctx = contextBuilder(sound.id)
      mockGetBySoundId.mockReturnValue(sound)

      await playerController.playSound(ctx)

      expect(ctx).toMatchSnapshot()
      expect(logger.debug).toMatchSnapshot()
      expect(playerService.playFile).toBeCalledWith(sound.filename)
    })
  })
})
