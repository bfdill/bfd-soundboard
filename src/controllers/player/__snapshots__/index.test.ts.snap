// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controllers -> player has known exports 1`] = `
Array [
  "PlayerController",
  "playerRouter",
]
`;

exports[`controllers -> player playSound bad request on non-string 1`] = `
Object {
  "body": Object {
    "soundId": undefined,
  },
  "params": Object {
    "soundId": undefined,
  },
  "status": 400,
}
`;

exports[`controllers -> player playSound bad request on non-string 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "message": "playSound().soundId !== string (undefined)",
        "soundId": undefined,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`controllers -> player playSound bad request on non-string 3`] = `
Object {
  "body": Object {
    "soundId": null,
  },
  "params": Object {
    "soundId": null,
  },
  "status": 400,
}
`;

exports[`controllers -> player playSound bad request on non-string 4`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "message": "playSound().soundId !== string (null)",
        "soundId": null,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`controllers -> player playSound bad request on non-string 5`] = `
Object {
  "body": Object {
    "soundId": false,
  },
  "params": Object {
    "soundId": false,
  },
  "status": 400,
}
`;

exports[`controllers -> player playSound bad request on non-string 6`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "message": "playSound().soundId !== string (false)",
        "soundId": false,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`controllers -> player playSound bad request on non-string 7`] = `
Object {
  "body": Object {
    "soundId": true,
  },
  "params": Object {
    "soundId": true,
  },
  "status": 400,
}
`;

exports[`controllers -> player playSound bad request on non-string 8`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "message": "playSound().soundId !== string (true)",
        "soundId": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`controllers -> player playSound bad request on non-string 9`] = `
Object {
  "body": Object {
    "soundId": 1,
  },
  "params": Object {
    "soundId": 1,
  },
  "status": 400,
}
`;

exports[`controllers -> player playSound bad request on non-string 10`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "message": "playSound().soundId !== string (1)",
        "soundId": 1,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`controllers -> player playSound not found on unknown song 1`] = `
Object {
  "body": Object {
    "soundId": "unknown",
  },
  "params": Object {
    "soundId": "unknown",
  },
  "status": 404,
}
`;

exports[`controllers -> player playSound not found on unknown song 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "message": "playSound().sound === undefined",
        "soundId": "unknown",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`controllers -> player playSound plays sound 1`] = `
Object {
  "body": Object {
    "sound": Object {
      "basename": "mommy shark",
      "filename": "daddy shark",
      "id": "baby_shark",
    },
  },
  "params": Object {
    "soundId": "baby_shark",
  },
  "status": 200,
}
`;

exports[`controllers -> player playSound plays sound 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "message": "playSound()",
        "sound": Object {
          "basename": "mommy shark",
          "filename": "daddy shark",
          "id": "baby_shark",
        },
        "soundId": "baby_shark",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
